btree/btree.c

```diff
@@ -2921,6 +2925,493 @@ static void ff_capture(struct slot *slot, struct m0_be_tx *tx)
 	M0_BTREE_TX_CAPTURE(tx, seg, h, hsize);
 }
 
+/**
+ *  --------------------------------------------
+ *  Section END - Fixed Format Node Structure
+ *  --------------------------------------------
+ */
+
+/**
+ *  --------------------------------------------
+ *  Section START -
+ *  Variable Sized Key and Values Node Structure
+ *  --------------------------------------------
+ *
+ * Proposed node structure
+ *
+ *
+ *                        +--------------------------+
+ *                        |                          |
+ *                 +------+-----------------------+  |
+ *                 |      |                       |  |
+ *             +---+------+-------------------+   |  |
+ *             |   |      |                   |   |  | DIR                    16   11       3  0
+ * +----------+v-+-v--+---v----+-------------++-+-++-++--+--+--+--------------+----+--------+--+
+ * |          |  |    |        |             |  |  |  |  |  |  |              |    |        |  |
+ * | Node Hdr |K0| K1 |   K3   |             +--+--+--+--+--+--+              | V3 |   V1   |V0|
+ * |          |  |    |        |             |  |  |  |  |  |  |              |    |        |  |
+ * +----------+--+----+--------+-------------++-++-+-++--+--+--+--------------+--^-+-----^--+-^+
+ *            0  5    12                      |  |   |                           |       |    |
+ *                                            +--+---+---------------------------+-------+----+
+ *                                               |   |                           |       |
+ *                                               +---+---------------------------+-------+
+ *                                                   |                           |
+ *                                                   +---------------------------+
+ *
+ *
+ *  The above structure represents the way variable sized keys and values may be
+ *  stored in memory.
+ *  Node Hdr or Node Header will store all the relevant info regarding this node
+ *  type.
+ *  The keys wil be added from the start of the node in an increasing way
+ *  whereas the values will be added from the end of the node in a decreasing 
+ *  way.
+ *  Additionally, we will maintain a directory in the central region of the node
+ *  which will contain the offsets to identify the address of the particular
+ *  key and value.
+ *
+ */
+#if 0
+struct dir_rec {
+	uint8_t key_offset;
+	uint8_t val_offset;
+};
+
+struct vkv_head {
+	struct m0_format_header  vkv_fmt;      /*< Node Header */
+	struct node_header       vkv_seg;      /*< Node type information */
+	uint16_t                 vkv_used;     /*< Count of records */
+	uint8_t                  vkv_shift;    /*< Node size as pow-of-2 */
+	uint8_t                  vkv_level;    /*< Level in Btree */
+	struct dir_rec          *dir_address;  /*< starting address of dir */
```

Comment:
We should keep this as a 32-bit variable by maintaining the offset of directory from the start of the node. This way we can have the vkv_used, vkv_shift, vkv_level and dir_address in one 64-bit space.

Reply:
Updated